### ISSUE TITLE
[Question] is request filter supported?

### ISSUE THREAD

**Issue by nchiring** [Skullabs/kikaha#145]
Kikaha looks a promusing micro framwork! I couldn't figure out from the doc and code base whether filter is supported.  If not yet, do you have any plan?

**Comment by miere** [comment]
Hi @nchiring, thanks for the feedback, and thanks for your question.

When you ask me for Filter support do you mean something like Servlet Filter which you can 'intercept' request and take some action before the actual request is done?

If your answer is "yes", than my answer is also "yes" ;) Sorry, may be I should make it more visible at documentation... Above I'll try to clarify how you can apply this filter strategy you asked me. If your answer was "no", please fill free to ask me again!

Kikaha is tiny layer above the amazing (but really low level) Undertow.io engine. it means that, under the hood, any request that Kikaha receives is handled by a io.undertow.server.HttpHandler implementation. Just as the Serlvet [Filter](https://docs.oracle.com/javaee/6/api/javax/servlet/Filter.html), this interface was designed to behaves just like a Responsibility Chain.

My suggestion to you is apply the following strategy to mimic the Filter pattern:

1. Implements your FilterHttpHandler
```java
import javax.inject.Singleton;
import io.undertow.server.*;

@Singleton
public class FilterHttpHandler implements HttpHandler {

  final HttpHandler executionChain;

  public FilterHttpHandler( HttpHandler executionChain ){
    this.executionChain = executionChain;
  }

  @Override
  public void handleRequest(final HttpServerExchange exchange) throws Exception {
    if ( /* here may be a rule that will 'filter' your request ... */ ) {
      // remember that Undertow is a Http Low Level engine, it may be verbose but is is easy to use...
      // Kikaha 2.0 (planned for July of this year) will have some helpers classes that will make this easier and less verbose...
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
      exchange.getResponseSender().send("Hello World");
    } else {
      // let the execution chain take its own way ;)
      executionChain.handleRequest( exchange );
    }
  }
}
```

Next, you should deploy HttpHandler as the root Handler using a deployment listener...
```java
import javax.inject.Singleton;
import kikaha.core.api.*;

@Singleton
public class FilterDeploymentListener implements DeploymentListener {

  @Override
  public void onDeploy( DeploymentContext context ) {
    HttpHandler currentRootHandler = context.rootHandler();
    HttpHandler filter = new FilterHttpHandler( currentRootHandler );
    context.rootHandler( filter );
  }

  @Override
  public void onUndeploy( DeploymentContext context ) {}
}
```
This implementation will guarantee that any request that your app receives will be delegated to your filter.

If you face any issue apply the above suggestion, please fill free to ask more questions here.

**Comment by nchiring** [comment]
Hey,
Thank you for the quick detail reply. Yes, I meant the same as you have understood.
This approach will work. However, for more conveniently is there any plan to support annotation based filter so that i can directly annotate my Rest handler methods/classes ala https://jersey.java.net/documentation/latest/filters-and-interceptors.html ? Say, I want to apply the filter for rest api /api/abc but not on /api/xyz, how can I achieve that? Sorry, I have just started exploring undertow APIs and so not having much idea about it.

**Comment by miere** [comment]
Nice questions, dude!! Let me try to answer them.

> I want to apply the filter for rest api /api/abc but not on /api/xyz, how can I achieve that?

Using the Undertow API you can easily identify which is the current request though the ```HttpServerExchange.getRelativePath``` method. I've changed my previous sample code to show you how you can identify the current request's URL. Also, I've also updated my previous answer to on how to set the executionChain in order to keep this handler Singleton (as you see on my note on the previous answer, sorry for this inconvenient).

```java
import javax.inject.Singleton;
import io.undertow.server.*;

@Singleton
public class FilterHttpHandler implements HttpHandler {

  HttpHandler executionChain;

  public FilterHttpHandler setExecutionChain( HttpHandler executionChain ){
    this.executionChain = executionChain;
  }

  @Override
  public void handleRequest(final HttpServerExchange exchange) throws Exception {
    if ( "/api/abc".equals( exchange.getRelativePath() ) ) {
      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
      exchange.getResponseSender().send("Hello World");
    } else {
      // let the execution chain take its own way ;)
      executionChain.handleRequest( exchange );
    }
  }
}
```
As you can see at the ```if ( "/api/abc".equals( exchange.getRelativePath() ) ) {``` line, is quite simple to detect which is the current request URL. In other hand, things like Headers, Cookies, Query and Path Parameters isn't that simple to retrieve from Undertow's API, especially if you need to convert them to a more convenient object other than String like "Integer", "Long", "Boolean" etc... Kikaha has some special services that could helps you to handle this. 

Bellow is the same example as above, but using a more convenient API provided by Kikaha.
```java
import javax.inject.*;
import io.undertow.server.*;

@Singleton
public class FilterHttpHandler implements HttpHandler {

  // You can use this object to write responses with "Response" objects
  @Inject ResponseWriter responseWriter;

  // You can use this object to retrieve data from the request
  @Inject RoutingMethodDataProvider methodDataProvider;

  // You can use this object to handle exceptions
  @Inject RoutingMethodExceptionHandler exceptionHandler;

  HttpHandler executionChain;

  public FilterHttpHandler setExecutionChain( HttpHandler executionChain ){
    this.executionChain = executionChain;
  }

  @Override
  public void handleRequest(final HttpServerExchange exchange) throws Exception {
    if ( shouldApplyFilter( exchange ) ) {
      Response response =  DefaultResponse.ok( "Hello World" ).contentType( Mimes.PLAIN_TEXT );
      responseWriter.write( exchange, response );
    } else {
      // let the execution chain take its own way ;)
      executionChain.handleRequest( exchange );
    }
  }

  /**
   * Will return true if the url is /api/abc and the query parameter named postId is greater than 1000
   */
  boolean shouldApplyFilter( final HttpServerExchange exchange ) {
    long postId = methodDataProvider.getQueryParam( exchange, "postId", Long.class );
    return "/api/abc".equals( exchange.getRelativePath() ) && postId > 1000;
  }
}
```
As you can see, despite the fact that Helper classes are separate services, it still easy to achieve your filter goals with this API. I'll include at the 2.0.0 version a centralized service that wrap both these three services together in order to be easier to handle requests through the Undertow API.

> However, for more conveniently is there any plan to support annotation based filter so that i can directly annotate my Rest handler methods/classes ala https://jersey.java.net/documentation/latest/filters-and-interceptors.html ?

I agree with you that it could be a little verbose use a DeploymentListener to deploy "filters". At this time, you should still use my earlier suggested DeploymentListener to deploy this filter. I will open a _feature request_ issue to include a better way to deploy filters at the 2.0.0 version, okay?

But, about the implementation, I'm not quite sure to do this through annotations. It is just a matter of available time to implement such feature on time to release the 2.0.0 version on July 1st. May be the annotation support will come at the 2.1.0 version (planned to November 1st), and at this release allow developers to configure the filter through a configuration file, what do you think?

**Comment by nchiring** [comment]
👍 I really appreciate not only your wonderful answer, but also your promptness in opening up tickets to capture the requirements ! I totally agree about implementation always needs not be annotation based. It can be any centralised/convenient approach which frees one from handling things on case by case basis.  Will keep an eye on these tickets. Subscribed to watch :)

**Comment by miere** [comment]
@nchiring Once I've finished both #147 and #148 issues, you will be able to use Filters in a (or quite similar to) Servlet way. Bellow the same example I show you before, but using the new API.

```java
import kikaha.core.modules.smart.*;
import kikaha.urouting.*;
import javax.inject.*;

@Singleton
// here is the simplified Filter implementation you suggested me (issue #148)
public class MyInterceptionFilter implements Filter {

  @Inject UndertowHelper undertow;

  public void doFilter(HttpServerExchange exchange, FilterChain chain) {
    // and here is the simplified Undertow API you suggested me (issue #147)
    SimpleExchange simpleExchange = undertow.simplify( exchange );

    // you can retrieve values in a very similar way the example I've suggested before,
    // but it is a lil' less verbose now
    long postId = simpleExchange.getQueryParameter( "postId", Long.class );

    if ( "/api/abc".equals( simpleExchange.getRelativePath() ) && postId > 1000 )
      simpleExchange.sendResponse( DefaultResponse.ok( "Hello World" ) );
    else
      chain.runNext();
  }
}
```

As you can see, even with the comments, it has way less lines of code than the previous approach.
Thank you for your feedback and suggestions. It will be available on all versions since the 2.0.0-beta4.
